### PR TITLE
test: optimize waitForNextUpstreamRequest()

### PR DESCRIPTION
**Pull request by lambdai**
Commit Message:
//test/integration:protocol_integration_test
After:
  Stats over 50 runs: max = 11.8s, min = 1.2s, avg = 4.0s, dev = 4.3s
Before:
  Stats over 50 runs: max = 2.6s, min = 1.4s, avg = 2.4s, dev = 0.3s

Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Additional Description:
Risk Level: LOW
Testing: test
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: test/integration/http_integration.cc [hunk 1]
```diff
@@ -357,16 +357,24 @@ HttpIntegrationTest::waitForNextUpstreamRequest(const std::vector<uint64_t>& ups
   absl::optional<uint64_t> upstream_with_request;
   // If there is no upstream connection, wait for it to be established.
   if (!fake_upstream_connection_) {
-
     AssertionResult result = AssertionFailure();
-    for (auto upstream_index : upstream_indices) {
+    int upstream_index = 0;
```
Comment (zuercher) [inline test/integration/http_integration.cc:361]: I wonder if the test failures are related to the fact that the original only tests upstreams with the given indices, while your version always checks all upstreams? Also, the original is much more likely to return the first member of upstream_indices with a request than the optimized version.

Comment (lambdai) [reply test/integration/http_integration.cc:361]: Thanks! I think you are right. I should iterate over the indexes which is passed. An embarrassing mistake.

File: test/integration/http_integration.cc [hunk 2]
```diff
@@ -357,16 +357,24 @@ HttpIntegrationTest::waitForNextUpstreamRequest(const std::vector<uint64_t>& ups
...
-
     AssertionResult result = AssertionFailure();
-    for (auto upstream_index : upstream_indices) {
-      result = fake_upstreams_[upstream_index]->waitForHttpConnection(
-          *dispatcher_, fake_upstream_connection_, connection_wait_timeout, max_request_headers_kb_,
-          max_request_headers_count_);
+    int upstream_index = 0;
+    Event::TestTimeSystem& time_system = timeSystem();
+    auto end_time = time_system.monotonicTime() + connection_wait_timeout;
+    // Loop over the upstreams until timeout or get a upstream request.
```
Comment (alyssawilk) [inline test/integration/http_integration.cc:364]: rephrase?
Loop over the upstreams until the call times out or an upstream request is received.

Comment (lambdai) [reply test/integration/http_integration.cc]: Updated!

**Pull request merged**